File: attendee.py
from tabulate import tabulate
from matplotlib import pyplot as plt
def event_list():
    print("List of events")
    if event==[]:
        print("No events yet")
    else:
        print(tabulate(event,headers="firstrow",tablefmt="grid"))

global event
event=[["S.No","Event Name","Event Date"],[1,"Python Workshop","20/11/2024"]]
attendee=[["Kushagra Chandra",'kc@gmailcom','Non-Vegetarian','False','Python Workshop']]
event_list()
event_choice=input("Want to add more events(Y/N):")
event_choice=event_choice.lower()
if event_choice=='y':
    print()
    event_name=input("Enter event name:")
    event_date=input("Enter event date(DD/MM/YYYY):")
    m=event[-1][0]
    event.append([m+1,event_name,event_date])
    print(event)
    print()
    event_list()

event_selection=int(input("Enter S.No of the event you want to see details of: "))
print('Event selected',event[event_selection][1])

event_attendee=[['S.No','Name','Email ID','Preference','Attended']]
#MENU
while True:
    print()
    ch=int(input('Enter 1 to see attendee details\nEnter 2 to add attendee\nEnter 3 to mark attendance\nEnter 4 to search for attendee\nEnter 5 to get event summary\nEnter 6 to exit form the menu: '))
    print()
    if ch==1:
        event_attendee=[['S.No','Name','Email ID','Preference','Attended']]
        j=1
        for i in attendee:
            if i[-1]==event[event_selection][1]:
                event_attendee.append(i[0:-1])
                event_attendee[j].insert(0,len(event_attendee)-1)
                j+=1
        if len(event_attendee)!=1:
            print(tabulate(event_attendee,headers='firstrow',tablefmt='grid'))
        else:
            print("No Attendees in the event")
    
    elif ch==2:
        print()
        naam=input('Enter Attendee Name: ')
        per=input('Enter Attendee Preference(Veg/Non-Veg): ')
        email=input('Enter Attendee mail id: ')
        attendance=input('Present in event or not(Y/N): ')
        attendance=attendance.lower()

        if attendance=='y':
            attendee.append([naam,email,per,'True',event[event_selection][1]])
        else:
            attendee.append([naam,email,per,'False',event[event_selection][1]])
    
    elif ch==3:
        x=1
        print('List of Attendees not present:')
        abs=[['S.No','Name','Email ID','Preference','Attended']]
        for i in attendee:
            if i[-1]==event[event_selection][1] and i[-2]=='False':
                abs.append(i[0:-1])
                abs[x].insert(0,len(abs)-1)
                x+=1
        print(tabulate(abs,headers='firstrow',tablefmt='grid'))
        while True:
            att_sl=int(input('\nType Serial No. of the Attendee to mark present(0 to exit): '))
            if att_sl==0:
                break
            else:
                for i in attendee:
                    if i[0]==abs[att_sl][1] and i[-1]==event[event_selection][1]:
                        i[-2]='True'
                        print('Attendance marked for',abs[att_sl][1])
    elif ch==4:
        search=input('Enter Attendee name you want to search(Pls write correct spelling): ')
        found=False
        for i in attendee:
            if i[0].lower().startswith(search.lower()):
                if i[-1]==event[event_selection][1]:
                    print("Attendee",search,' is found')
                    det=input('Do you want to know the details(Y/N): ')
                    if det.lower()=='y':
                        details=[['Name','Email','Preference','Attended','Event']]
                        details.append(i)
                        print(tabulate(details,headers='firstrow',tablefmt='grid'))
                    found=True
                    break
        if not found:
            print('Attendee',search,'not found')
    elif ch==5:
        count=0
        n=0
        v=0
        p=0
        for i in attendee:
            if i[-1]==event[event_selection][1]:
                count+=1
                if i[-2]=='True':
                    p+=1
                if i[2].lower().startswith('non'):
                    n+=1
                else:
                    v+=1
        x=['Total No. of Attendee','Veg Preference','Non-Veg Preference','Present in Event']
        y=[count,v,n,p]
        plt.figure(figsize=(8, 6))
        plt.bar(x,y)
        plt.show() 
        print('Total No. of attendees in the Event',event[event_selection][1],'are:',count)
        print('No. of attendees with Veg preference are:',v)
        print('No. of attendees with Non-Veg preference are:',n)
        print('Attendance percentage:',(p/count)*100,'%')
        
    elif ch==6:
        print('Thank you for using our program\nHave a nice day :)')
        break
                
            

File: attendee.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_EVENTS 100
#define MAX_ATTENDEES 500
#define MAX_NAME_LEN 50
#define MAX_EMAIL_LEN 100
#define MAX_DATE_LEN 20

typedef struct {
    int serial_no;
    char name[MAX_NAME_LEN];
    char date[MAX_DATE_LEN];
} Event;

typedef struct {
    char name[MAX_NAME_LEN];
    char email[MAX_EMAIL_LEN];
    char preference[MAX_NAME_LEN];
    int attended; // 1 for true, 0 for false
    int event_serial_no;
} Attendee;

Event events[MAX_EVENTS];
Attendee attendees[MAX_ATTENDEES];
int event_count = 1;
int attendee_count = 0;

void list_events() {
    printf("List of events:\n");
    if (event_count == 1) {
        printf("No events yet.\n");
    } else {
        printf("%-5s %-30s %-15s\n", "S.No", "Event Name", "Event Date");
        for (int i = 0; i < event_count - 1; i++) {
            printf("%-5d %-30s %-15s\n", events[i].serial_no, events[i].name, events[i].date);
        }
    }
}

void add_event() {
    printf("Enter event name: ");
    getchar(); // Clear the newline buffer
    fgets(events[event_count - 1].name, MAX_NAME_LEN, stdin);
    events[event_count - 1].name[strcspn(events[event_count - 1].name, "\n")] = 0; // Remove newline

    printf("Enter event date (DD/MM/YYYY): ");
    fgets(events[event_count - 1].date, MAX_DATE_LEN, stdin);
    events[event_count - 1].date[strcspn(events[event_count - 1].date, "\n")] = 0; // Remove newline

    events[event_count - 1].serial_no = event_count;
    event_count++;
}

void view_event_details(int serial_no) {
    printf("Event selected: %s\n", events[serial_no - 1].name);
}

void view_attendees(int event_serial_no) {
    printf("%-5s %-30s %-30s %-15s %-10s\n", "S.No", "Name", "Email ID", "Preference", "Attended");
    int count = 0;
    for (int i = 0; i < attendee_count; i++) {
        if (attendees[i].event_serial_no == event_serial_no) {
            printf("%-5d %-30s %-30s %-15s %-10s\n", 
                i + 1, attendees[i].name, attendees[i].email, 
                attendees[i].preference, attendees[i].attended ? "Yes" : "No");
            count++;
        }
    }
    if (count == 0) {
        printf("No attendees for this event.\n");
    }
}

void add_attendee(int event_serial_no) {
    printf("Enter Attendee Name: ");
    getchar();
    fgets(attendees[attendee_count].name, MAX_NAME_LEN, stdin);
    attendees[attendee_count].name[strcspn(attendees[attendee_count].name, "\n")] = 0;

    printf("Enter Attendee Email: ");
    fgets(attendees[attendee_count].email, MAX_EMAIL_LEN, stdin);
    attendees[attendee_count].email[strcspn(attendees[attendee_count].email, "\n")] = 0;

    printf("Enter Attendee Preference (Veg/Non-Veg): ");
    fgets(attendees[attendee_count].preference, MAX_NAME_LEN, stdin);
    attendees[attendee_count].preference[strcspn(attendees[attendee_count].preference, "\n")] = 0;

    printf("Is the attendee present (Y/N): ");
    char choice;
    scanf(" %c", &choice);
    attendees[attendee_count].attended = (choice == 'Y' || choice == 'y') ? 1 : 0;

    attendees[attendee_count].event_serial_no = event_serial_no;
    attendee_count++;
}

void mark_attendance(int event_serial_no) {
    printf("List of attendees not present:\n");
    for (int i = 0; i < attendee_count; i++) {
        if (attendees[i].event_serial_no == event_serial_no && !attendees[i].attended) {
            printf("%-5d %-30s\n", i + 1, attendees[i].name);
        }
    }

    while (1) {
        printf("Enter S.No of attendee to mark present (0 to stop): ");
        int serial_no;
        scanf("%d", &serial_no);
        if (serial_no == 0) {
            break;
        } else if (serial_no > 0 && serial_no <= attendee_count) {
            attendees[serial_no - 1].attended = 1;
            printf("Attendance marked for %s\n", attendees[serial_no - 1].name);
        } else {
            printf("Invalid serial number.\n");
        }
    }
}

void search_attendee(int event_serial_no) {
    char search_name[MAX_NAME_LEN];
    printf("Enter the name of the attendee to search: ");
    getchar();
    fgets(search_name, MAX_NAME_LEN, stdin);
    search_name[strcspn(search_name, "\n")] = 0;

    for (int i = 0; i < attendee_count; i++) {
        if (attendees[i].event_serial_no == event_serial_no &&
            strstr(attendees[i].name, search_name) != NULL) {
            printf("Attendee found: %s\n", attendees[i].name);
            return;
        }
    }
    printf("Attendee not found.\n");
}

void event_summary(int event_serial_no) {
    int total = 0, veg = 0, non_veg = 0, present = 0;
    for (int i = 0; i < attendee_count; i++) {
        if (attendees[i].event_serial_no == event_serial_no) {
            total++;
            if (strstr(attendees[i].preference, "Non")) {
                non_veg++;
            } else {
                veg++;
            }
            if (attendees[i].attended) {
                present++;
            }
        }
    }

    printf("Event Summary:\n");
    printf("Total attendees: %d\n", total);
    printf("Veg attendees: %d\n", veg);
    printf("Non-Veg attendees: %d\n", non_veg);
    printf("Attendance percentage: %.2f%%\n", total ? (present / (float)total) * 100 : 0);
}

int main() {
    // Predefined event and attendee
    strcpy(events[0].name, "Python Workshop");
    strcpy(events[0].date, "20/11/2024");
    events[0].serial_no = 1;

    strcpy(attendees[0].name, "Kushagra Chandra");
    strcpy(attendees[0].email, "kc@gmailcom");
    strcpy(attendees[0].preference, "Non-Vegetarian");
    attendees[0].attended = 0;
    attendees[0].event_serial_no = 1;
    attendee_count++;
    event_count++;

    list_events();

    int choice;
    while (1) {
        printf("\n1. List events\n2. Add event\n3. View event details\n");
        printf("4. View attendees\n5. Add attendee\n6. Mark attendance\n");
        printf("7. Search attendee\n8. Event summary\n9. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        if (choice == 1) {
            list_events();
        } else if (choice == 2) {
            add_event();
        } else if (choice == 3) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            view_event_details(serial_no);
        } else if (choice == 4) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            view_attendees(serial_no);
        } else if (choice == 5) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            add_attendee(serial_no);
        } else if (choice == 6) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            mark_attendance(serial_no);
        } else if (choice == 7) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            search_attendee(serial_no);
        } else if (choice == 8) {
            int serial_no;
            printf("Enter event serial number: ");
            scanf("%d", &serial_no);
            event_summary(serial_no);
        } else if (choice == 9) {
            printf("Thank you for using the program. Goodbye!\n");
            break;
        } else {
            printf("Invalid choice. Try again.\n");
        }
    }

    return 0;
}